Coalesce poll events on queue eviction

vortigont opened the pull request:
Coalesce poll events on queue eviction

Refer: mathieucarbou/ESPAsyncWebServer#165

Try to coalesce two (or more) consecutive poll events into one.
This usually happens with poor implemented user-callbacks that are runs too long and makes poll events to stack in the queue if consecutive user callback for a same connection runs longer that poll time then it will fill the queue with events until it deadlocks.
This is a workaround to mitigate such poor designs and won't let other events/connections to starve the task time.
It won't be effective if user would run multiple simultaneous long running callbacks due to message interleaving.

**queue deadlock and congestion avoidance**

 - do not let `_remove_events_with_arg()` call to block on queue pertubation, otherwise it could deadlock

 - in any case do not block on adding LWIP_TCP_POLL event to the queue, it does not make much sense anyway
   due to poll events are repetitive

 - `_get_async_event()` will discard LWIP_TCP_POLL events when q gets filled up
   this will work in combination with throttling poll events when adding to the queue
   Poor designed apps and multiple parallel connections could flood the queue with interleaved poll events
   that can't be properly throttled or coalesced. So we can discard it in the eviction task after coalescing
   It will work in this way:
    - queue is up to 1/4 full - all events are entering the queue and serviced
    - queue is from 1/4 and up to 3/4 full - new poll events are throttled on enqueue with linear-increasing probability
    - queue is from 3/4 up to full top - all new poll events are ignored on enqueue and existing poll events
      already in the queue are discarded on eviction with linear-increasing probability giving away priority for all other
      events to be serviced. It is expected that on a new poll timer connection polls could reenter the queue

mathieucarbou commented:
Thanks! Let me know when this is ready to review, I will do some perf tests.

mathieucarbou commented on src/AsyncTCP.cpp:
```diff
@@ -226,15 +259,23 @@ static void _async_service_task(void* pvParameters) {
+      switch((size_t)packet->arg){
+        case LWIP_TCP_RECV:
+        case LWIP_TCP_POLL:
+          if (esp_task_wdt_add(NULL) != ESP_OK) {
+            log_e("Failed to add async task to WDT");
+          }
+          _handle_async_event(packet);
+          if (esp_task_wdt_delete(NULL) != ESP_OK) {
```
There are things weird related to the TWDT:
- It is only activated if `CONFIG_ASYNC_TCP_RUNNING_CORE` is not set. If `CONFIG_ASYNC_TCP_RUNNING_CORE=1` is set (i.e. what I use to do), then there is no watchdog.
- If `CONFIG_ASYNC_TCP_USE_WDT` is set, `CONFIG_ASYNC_TCP_USE_WDT=1` then can to be set if a watchdog is active on ore 1

All that to say that by default, TWDT is activated and otherwise it is user-controled.

So should the calls to `esp_task_wdt_add` and `esp_task_wdt_delete` be only around the use callbacks instead of here wrapping also the AsyncTCP code ?

I don't know how costly it is to call `esp_task_wdt_add` and `esp_task_wdt_delete` frequently, but it seems quite inefficient... These things should only be there to protect against long user callbacks right ?

mathieucarbou replied:
`LWIP_TCP_POLL ` => goes to onPolll callback, which is controlled by ESpAsyncWS internally and dos not leak to the user

`LWIP_TCP_RECV ` => onData => this is the one that could trigger some user code callback.

So IMO, we could just surround this part of code with `esp_task_wdt_add` and `esp_task_wdt_delete ` ?

```
      if (_recv_cb) {
        _recv_cb(_recv_cb_arg, this, b->payload, b->len);
      }
```

vortigont replied:
yeah, I do not like this enable/disable either. Better I will rework it in a bit different way. And also clean up this define mess

vortigont commented:
I reworked it.
Now WDT could just enabled or disabled despite of the core affinity. If enabled it would constantly guard the asycntcp task with periodic feeds on each event.
Actually with much improved queue control I'm no longer sure that WD could be of much use by default.

Also added more safeguards on queue operation. This should make a big difference in robustness and congestion handling.

mathieucarbou commented:
> I reworked it. Now WDT could just enabled or disabled despite of the core affinity. If enabled it would constantly guard the asycntcp task with periodic feeds on each event. Actually with much improved queue control I'm no longer sure that WD could be of much use by default.
> 
> Also added more safeguards on queue operation. This should make a big difference in robustness and congestion handling.

I think the WDT is mostly useful for the user ondata callback to make sure the user code runs fast and does not do some delay or long loop tasks. I think it has a value and should be activated by default.

mathieucarbou commented:
moving to draft until testing is done (see https://github.com/mathieucarbou/ESPAsyncWebServer/pull/171)
